src/dnstest_harness.erl

```diff
@@ -188,15 +217,131 @@ test_records(ExpectedRecords, ActualRecords, SectionType) ->
         true ->
-            true
+            ?LOG_INFO("Record count mismatch in ~p section.", [SectionType]),
+            ?LOG_INFO("Expected (~p): ~p", [length(ExpectedRecordsSorted), ExpectedRecordsSorted]),
+            ?LOG_INFO("Actual\s\s\s(~p): ~p", [length(ActualRecordsSorted), ActualRecordsSorted]),
+            false;
+        false ->
+            ZippedRecords = lists:zip(ExpectedRecordsSorted, ActualRecordsSorted),
+            lists:foldl(
+                fun({Expected, Actual}, Acc) ->
+                    case Expected =:= Actual of
+                        true ->
+                            % Continue if they match
+                            Acc;
+                        false ->
+                            % Log detailed differences
+                            compare_record_fields(Expected, Actual, SectionType),
+                            % Mismatch found, overall result is false
+                            false
+                    end
+                end,
+                % Initial accumulator value (true means no mismatch found yet)
+                true,
+                ZippedRecords
+            )
     end.
 
+% Helper to compare fields of mismatched record tuples
+-spec compare_record_fields(
+    {dns:dname(), dns:class(), dns:type(), dns:ttl(), dns:rrdata()},
+    {dns:dname(), dns:class(), dns:type(), dns:ttl(), dns:rrdata()},
+    section()
+) -> ok.
+compare_record_fields(
+    {NameE, ClassE, TypeE, TTLE, DataE}, {NameA, ClassA, TypeA, TTLA, DataA}, SectionType
+) ->
+    ?LOG_INFO("Record mismatch in ~p section:", [SectionType]),
+    ?LOG_INFO_PAD(2, "Full Expected: ~p", [{NameE, ClassE, TypeE, TTLE, DataE}]),
+    ?LOG_INFO_PAD(2, "Full Actual:\s\s\s~p", [{NameA, ClassA, TypeA, TTLA, DataA}]),
+    case NameE =:= NameA of
+        false ->
+            ?LOG_INFO_PAD(4, "Field 'Name' mismatch: Expected ~p, Actual ~p", [NameE, NameA]);
+        true ->
+            ok
+    end,
+    case ClassE =:= ClassA of
+        false ->
+            ?LOG_INFO_PAD(4, "Field 'Class' mismatch: Expected ~p, Actual ~p", [ClassE, ClassA]);
+        true ->
+            ok
+    end,
+    case TypeE =:= TypeA of
+        false ->
+            ?LOG_INFO_PAD(
+                4,
+                "Field 'Type' mismatch: Expected ~p (~s), Actual ~p (~s)",
+                [TypeE, dns:type_name(TypeE), TypeA, dns:type_name(TypeA)]
+            );
+        true ->
+            ok
+    end,
+    case TTLE =:= TTLA of
+        false ->
+            ?LOG_INFO_PAD(4, "Field 'TTL' mismatch: Expected ~p, Actual ~p", [TTLE, TTLA]);
+        true ->
+            ok
+    end,
+    case DataE =:= DataA of
+        false ->
+            ?LOG_INFO_PAD(4, "Field 'Data' mismatch: Expected ~p, Actual ~p", [DataE, DataA]),
+            % Add specific formatting for known binary-containing types
+            % Check TypeE first, assuming Type mismatch was already caught if they differ.
```

Comment:
I wouldn't have expected this much detail in the comparisons below here, but if it's useful for debugging tests, then excellent, thanks for doing it!